ui-spacetimechart/src/utils/canvas.ts

```diff
@@ -273,3 +273,39 @@ export function drawPathExtremity(
     drawPathStopExtremity(ctx, timePixel, spacePixel, swapAxis);
   }
 }
+
+/**
+ *
+ * @param minT number timestamp
+ * @param maxT number timestamp
+ * @param timeRanges time frames (24h, 12h, 6h, …)
+ * @param gridlinesLevels width of the lines for each time frame
+ * @returns Record<number, number>
+ * Keys are times in ms
+ * Values are the highest level on each time
+ */
+
+export function displayElementsBasedOnZoom<T>(
+  minT: number,
+  maxT: number,
+  timeRanges: number[],
+  levels: number[],
+  formatter: (level: number, i: number) => T = identity
+) {
+  const result: Record<number, T> = {};
+  const localOffset = -new Date(minT).getTimezoneOffset() * 60 * 1000;
```

Comment:
In general I'm not a fan of using `getTimezoneOffset(), because it behaves wrongly on timezone offset change. For instance if daylight savings are applied on the left side of the chart but aren't on the right side, the chart will be wrong.

I'd suggest applying the offset only to `minT` since it's the only timestamp it's valid for:

```ts
const localOffset = new Date(minT).getTimezoneOffset() * 60 * 1000;
let t = Math.floor((minT - localOffset) / range) * range + localOffset;
```

then the condition inside the loop can be dropped, and all other `localOffset` operations can be removed.

Reply:
that issue existed already, we have days with only 23h or 25h, everything becomes shifted, because we add 86400 s each time.

thanks, it’s neater like this.
Why do we need the offset inside the Math.floor ?
I think `let t = Math.floor(minT / range) * range - minTLocalOffset;` is enough.

Reply:
My reasoning is that all timestamps are Unix time without any offset. If we apply the offset once without undoing it, then it's not Unix time anymore, or am I missing something?

Maybe the `Math.floor()` makes it so we get the correct number by chance if we apply the offset only once? However I think it'd break on other `range` and offset values.

Reply:
We’re still getting a valid unix timestamp. But instead of getting the timestamp of 0h UTC we get the timestamp of 22h UTC (so that it show 0h Paris and not 2h Paris like before)

![image_50353153](https://github.com/user-attachments/assets/2016484b-2cba-4536-bcc9-a024ac33f851)

Reply:
Sorry, I still don't believe it's correct. `minT` is the earliest timestamp in the time window. We want to find the first graduation marker. We need to snap `minT` to a `range`-spaced grid shifted by the offset. Everything needs to be UTC apart from the snapping.

Your solution works as long as `minTLocalOffset` is negative, because the loop keeps going until it reaches `maxT` (it'll just run longer). However, it breaks with positive timezone offsets. You can try e.g. `minTLocalOffset = 6 * 60 * 60 * 1000` and notice missing grid lines on the left.

Reply:
Ok I see the issue, we get t snapped to 0h local time, but we get 0h the day+1, so we miss some of the range, thanks for digging that up ! 

I’ve written a test to take that case into account

Reply:
Sweet, thanks for adding a test!